### ISSUE TITLE
Simplify ArtifactReference fields?

### ISSUE THREAD

**Issue by marcelamelara** [slsa-framework/slsa#614]
Do we need both `uri` and `downloadLocation` fields when artifacts are always matched on `digest`? 

This question came out of a discussion happening over at https://github.com/in-toto/attestation/pull/133, so I wanted to bring this topic into the SLSA community. 

Currently the ArtifactReference description in the Provenance v1.0 draft says that the `uri` must be set when the `digest` is not, implying that matching can in fact occur on the `uri`. This seems to be in conflict with our current definition of "artifact", which is any object that is hashable.

My take is that, we might not need the `uri` field in an ArtifactReference. For one, it makes the model for artifact matching less ambiguous. It seems like the `uri` is being used as both a type and (in some cases) location hint of the artifact, but IMO these two aspects can be clearly indicated via the `mediaType` and `downloadLocation` in an ArtifactReference.

Would love to hear folks' thoughts on this. cc/ @MarkLodato @TomHennen

**Comment by di** [comment]
My two cents: an "artifact" implies something to be hashable.

In https://github.com/in-toto/attestation/pull/133#issuecomment-1422692712, an example of a URI for a builder is given as being something that could be a non-hashable `ArtifactReference`, but I think that would not be a common assumption for most people, and I agree with @marcelamelara's suggestion https://github.com/in-toto/attestation/pull/133#issuecomment-1425044269 that we should handle artifact and service references separately.

Given that, I think the concrete things we could do here in Provenance v1.0 is:

- make `digest` required for `ArtifactReference`;
- create a separate `ServiceReference` for un-hashable references;
- expand `builderDependencies` to permit both `ArtifactReference` and `ServiceReference` types.

That said, I think retaining the `uri` field for artifacts is worthwhile, as it provides additional metadata that might not otherwise be available via the `mediaType` or `downloadLocation`, like ecosystem, version, etc.

**Comment by marcelamelara** [comment]
Thanks for your input @di!

There's actually a third use case for references that aren't services, but more broadly non-hashable resources/entities, such as GitHub pull requests (see [ITE-4](https://github.com/in-toto/ITE/tree/master/ITE/4#use-case-2-github-entities)). In that case, maybe defining something like a `ResourceReference` would encapsulate all of these non-artifact use cases?

Through discussions around changing the `subject` field type in the in-toto Statement, we also started to discuss the potential for reference sub-types (https://github.com/in-toto/attestation/pull/133#discussion_r1107357377). This is a slightly different approach to distinguishing between references to artifacts vs non-hashable services/resources, but might help with some simplification in field types in predicates, since references to both artifacts and services non-hashable resources will likely still have some common fields (e.g., `localName`, `uri`, `annotations`). 

Whichever approach we choose, completely separate references types, or sub-types, I think the main goal is making sure that references to artifacts _always_ have a `digest` field.

> That said, I think retaining the `uri` field for artifacts is worthwhile, as it provides additional metadata that might not otherwise be available via the `mediaType` or `downloadLocation`, like ecosystem, version, etc.

Good point, and I think we've converged towards adding the `uri` field back into the reference type in in-toto.

**Comment by MarkLodato** [comment]
+1 to the term "resources" rather than "services".

Any suggestions for how to handle `resolvedDependencies`? Right now, it is a single array, like so:

```jsonc
"resolvedDependencies": [
  // A VM image that does have a digest, but the digest is unavailable to the provenance generator:
  {"uri": "vm_image:https://github.com/actions/virtual-environments/releases/tag/ubuntu20/20220515.1"},
  // A source repository that is not content addressable, thus computing the digest is infeasible:
  {"uri": "p4://depot/main@1234"},
  // A source repository that is content addressable:
  {"uri": "git+https://github.com/octocat/hello-world",
   "digest": {"sha1": "abcd1234..."}},
]
```

Would you make that two separate arrays, or something else?

**Comment by TomHennen** [comment]
The proposal to have ArtifactReferences and ResourceReferences sounds fine to me I suppose (though I'm not sure how much it will matter in practice).

If that route is taken I do think it would be good to make them separate fields in the json.

Either something like

```
"resolvedArtifactDeps": [
],
"resolvedResourceDeps":[
]
```

or

```
"resolvedDependencies": [
  {"artifact": <ArtifactReference>},
  {"resource":<ResourceReference>},
  ...
]

The alternative of 

```
"resolvedDependencies": [<ArtifactReference|ResourceReference>]
```

Would be very annoying.

**Comment by marcelamelara** [comment]
Agreed that having two type options for the same field would be confusing. I suppose there's a third option for `resolvedDependencies`:

```
"resolvedDependencies": {
  "artifacts": [<ArtifactReference>, ...],
  "resources": [<ResourceReference>, ...],
}
```

**Comment by di** [comment]
That makes sense to me! (As does `ResourceReference` instead of `ServiceReference`).

**Comment by MarkLodato** [comment]
Sorry, I still don't understand the reasoning for creating two different types. Could someone explain that? Won't both `ArtifactReference` and `ResourceReference` have exactly the same fields, except:
- `digest` is required for `ArtifactReference` and absent for `ResourceReference`
- `uri` is optional for `ArtifactReference` and required for `ResourceReference`

**Comment by TomHennen** [comment]
I think that's right.  Probably one of the reasons I'm ambivalent about making the distinction.  Can probably work just as well with a couple lines of documentation.

**Comment by marcelamelara** [comment]
My concern is the parsing and enforcement of the required fields on a consumer's end. With a single reference type, that becomes a little trickier. Are we going to say that if a `digest` is set, then the assumption is that it's an artifact? What is matched against the digest, the artifact located at the `uri` or `downloadLocation`? When both `uri` and `downloadLocation` are set, which takes precedence to match against `digest`? If only `uri` is set, but no `digest`, what is the reference matched on? 

I don't disagree that having a single type would be more convenient to manage from a spec and metadata producer's perspective. And, these sorts of questions have come up a few times in the past, albeit in a slightly different context, but with no clear resolution AFAICT (e.g. https://github.com/in-toto/attestation/issues/5#issuecomment-1061041809). I think the concerns discussed on the in-toto side apply as much to any other references that are being bound in the Provenance (or any in-toto predicate), so I just want to make sure we can converge on something.

**Comment by marcelamelara** [comment]
Appreciate the examples! 

How does your verifier know when to ignore and not to ignore the digest, whether or not it's set? Based on your description, it sounds like the verifier is doing some sort of analysis on the URI to decide whether to match that object to the digest. 

But this is exactly my point. Given a bunch of different URIs in a reference, how _should_ a consumer decide whether or not to match the digest as part of its policy? In the first example, the repo digest seems more informational because the object the URI points to is mutable, whereas in the second and third examples, those objects are (assumed to be) immutable, which is presumably why a matching step of the digest against the object at the URI is part of the policy. 

I know this by looking at the URIs (and possibly other info in the reference or other metadata), but unless we expect that all verifiers will perform a deeper analysis on each URI in a reference, how are they supposed to decide which type of policy should apply to the referenced thing?

Some way to distinguish artifacts (i.e. immutable objects) from resources (mutable objects) in the reference itself I think would facilitate this sort of policy assignment on the consumer's end. As a verifier, if I know the item in the reference is immutable, I SHOULD be able to match it against a specific digest.

**Comment by MarkLodato** [comment]
I don't think it's up to the producer to tell the consumer how to verify. Instead, the producer's job is to record what it did/knows: for example, it fetched a resource represented by `uri`, which resolved to `downloadLocation` and `digest`, and was interpreted as `mediaType`. In the case of `twoPartyReview`, the producer is saying, "I used some API to verify that this particular commit met the definition of `twoPartyReview`".

It's also not about mutable vs immutable. In my example, the first two were mutable references to immutable objects (i.e. it was a URL that points to a hashable thing, and that URL can change to point to something else) while the third was an immutable reference to an immutable object (the commit hash was in the URL, and the file's hash was the `digest`).

Here's a hypothetical example policy (made up syntax):

```json
{
  "requiredSlsaLevel": ["BUILD_L3"],
  "dependencyExceptions": [
    {
       "uri": "git+https://android.googlesource.com/something/else",
       "exception": "TWO_PARTY_REVIEW"
    },
    {
       "uri": "container_image://gcr.io/my-build-farm/farm1234",
       "exceptions": "BUILD_L0"
    }
  ]
}
```

Then the logic is something like:

```none
for dep in resolvedDependencies:
    if dep.uri is a source URI:
        if dep.uri has a 'TWO_PARTY_REVIEW' exception:
            pass
        else if dep.extensions.twoPartyReview:
            pass
        else if there exists a positive two-party review attestations for (dep.uri, dep.digest):
            pass
        else:
            raise an alert
    else:
        if dep.uri has a 'BUILD_L0' exception:
            pass
        else if there exists a VSA for (dep.uri, dep.digest) at policy.requiredSlsasLevel or higher:
            pass
        else:
            raise an alert
```

Does that help?

**Comment by marcelamelara** [comment]
Per our discussion at the SLSA Spec meeting today, there seems to be a broader discussion around how to verify `resolvedDependencies` coming out of this issue, and that falls more under the Source Track.

From a spec standpoint, just so we can resolve this on the in-toto spec side, it sounds like the requirements for references for are to identify artifacts based on their location _and_ digest, when available. Given your examples, I think a single `Reference` field type can cover most of those cases. But I also want to make sure that the `Reference` field conveys enough information to allow a consumer of Build Provenance to identify the referenced object with a high degree of confidence, given the various options of mutable/immutable references to immutable objects.

For me, what's still unclear is how a consumer of Build Provenance can know if a given reference (mutable or not) is pointing to an immutable object, given that the `digest` isn't currently required in an `ArtifactReference`. I hadn't been able to articulate this before, but I'm concerned about a case in which a compromised builder includes only a mutable reference `uri` and omits the `digest` for a dependency that should be immutable, in order to get a consumer to accept a package that was built using a malicious dependency. This is especially problematic if the consumer doesn't know the expected dependencies for a package a priori.

In a future where sources/dependencies are accompanied by Source Provenance/attestations, whether a referenced object is immutable could probably be indicated in the source attestation. But until then, or in cases where source information isn't available, I still think it would be valuable to be able to somehow explicitly indicate in a `Reference` field whether a referenced object is immutable.

In the examples we've been discussing, it seems like (1) using the `uri` field is quite typical, and (2) there's an implicit assumption that a `digest` will be provided when the reference points to an immutable object. So, would it be safe to say that the semantics of a reference that includes both `uri` and `digest` fields points to an immutable object? Can we make these requirements stronger for known-immutable objects?

**Comment by MarkLodato** [comment]
> I'm concerned about a case in which a compromised builder includes only a mutable reference `uri` and omits the `digest` for a dependency that should be immutable, in order to get a consumer to accept a package that was built using a malicious dependency

If the builder is compromised, isn't the entire provenance compromised? Under what scenario would an adversary be able to forge an ArtifactReference but not the rest of the provenance?

> So, would it be safe to say that the semantics of a reference that includes both `uri` and `digest` fields points to an immutable object? Can we make these requirements stronger for known-immutable objects?

I think it's a good idea to RECOMMEND populating both `uri` and `digest` when known. To make it REQUIRED, I suspect we'll need a higher SLSA Build level.

**Comment by MarkLodato** [comment]
I like "descriptor"! Good idea! That is much better than "reference". What do you think about `ResourceDescriptor`? That could resolve the "is it an artifact or not" question.

**Comment by di** [comment]
@marcelamelara That makes sense to me!

**Comment by marcelamelara** [comment]
I've opened PR https://github.com/in-toto/attestation/pull/143. Please let me know if this new specification for `ResourceDescriptors` meets everyone's requirements.

**Comment by MarkLodato** [comment]
Thanks, @marcelamelara! Once that is landed, I suggest that SLSA just reference that type.

**Comment by marcelamelara** [comment]
Addressed by #684
